src/convert.rs

```diff
@@ -424,132 +406,111 @@ pub fn elf_to_tbf(
-                    // check if there is any padding required.
-                    let padding = start - end;
-
-                    if padding < 1024 {
-                        if padding > 0 {
-                            if verbose {
-                                println!("  Adding {} bytes of padding between sections", padding,);
+    let mut start_address: u64 = 0;
+    let start_program_binary_index = binary_index;
+
+    // Iterate over ELF's Program Headers to assemble the binary image as a contiguous
+    // memory block. Only take into consideration segments where filesz is greater than 0
+    for segment in &elf_file.phdrs {
+        match segment.progtype {
+            elf::types::PT_LOAD => {
+                if segment.filesz > 0 {
+                    if last_segment_address_end.is_some() {
+                        // We have a previous section. Now, check if there is any
+                        // padding between the sections in the .elf.
+                        let padding = segment.paddr as usize - last_segment_address_end.unwrap();
+                        if padding < 1024 {
+                            if padding > 0 {
+                                if verbose {
+                                    println!("Padding between Program Segments of {}", padding);
+                                }
                             }
 
-                            // Increment our index pointer and add the padding bytes.
-                            binary_index += padding;
                             let zero_buf = [0_u8; 1024];
                             binary.extend(&zero_buf[..padding]);
+                            binary_index += padding;
+                        } else {
+                            println!(
+                                "Warning! Padding to Program segment is too large ({} bytes).",
+                                padding
+                            );
                         }
                     } else {
-                        println!(
-                            "Warning! Padding to section {} is too large ({} bytes).",
-                            section.shdr.name, padding
-                        );
+                        // This is the first segment, take the Physical Address as the starting
+                        // point to compute offsets from
+                        start_address = segment.paddr;
+
+                        // check if start of segment matches the fixed address flash (for non-PIC files)
+                        if let Some(fixed_start_address) = fixed_address_flash {
+                            if start_address != fixed_start_address as u64 {
+                                panic!("First Segment address ({:#x}) is different than fixed_address_flash ({:#x})",
+                                    start_address,
+                                    fixed_start_address);
+                            }
+                        }
```

Comment:
This check seems backwards. We figure out what the fixed flash address is from the elf, so it doesn't make sense to then check if we figured out the fixed flash address correctly.